### PR TITLE
get root authentication in a cleaner way

### LINKED ISSUE #2447: Suggestion on TicketGrantingTicketImpl
I continue having a look into TicketGrantingTicketImpl, trackServiceSession method:

```
final List<Authentication> authentications = getChainedAuthentications();
service.setPrincipal(authentications.get(authentications.size() - 1).getPrincipal());
```

as far as I can see, getChainedAuthentications() return a list where the root authentication is the last element.
So, the previous 2 lines could be replace by getRoot().getAuthentication(). Am I right?

### PR DESCRIPTION
**Pull request by DavidRG13**
Closes #2447 

Changes:
- get root authentication in a cleaner way

### CODE REVIEW

**Comment by DavidRG13** [comment]
Please add/suggest more test cases if needed.

Also, this was used in several places, would it be useful to create a new method in TicketGrantingTicket for this? ```getRootAuthentication()```

**Comment by mmoayyed** [comment]
> Also, this was used in several places, would it be useful to create a new method in TicketGrantingTicket for this? getRootAuthentication()

I think so, yes. perhaps under `TicketRegistrySupport`

**Comment by DavidRG13** [comment]
All classes implementing ```TicketGrantingTicket```  should implement ```getRoot()```, why not creting the new method there?

**Comment by mmoayyed** [comment]
Because:

1. Your proposal is mostly about adding a convenience method. ` ticketGrantingTicket.getRoot().getAuthentication()` -> ` ticketGrantingTicket.getRootAuthentication()`. This is not part of the API. Just something that makes it easier to work with the API. And certainly, we are not going to redecorate the API altogether because it saves a few keystrokes in this case.
2. API changes are not allowed until a major release in most cases.

File: core/cas-server-core-services/src/main/java/org/apereo/cas/services/RegisteredServiceAccessStrategyUtils.java [hunk 1]
```diff
@@ -127,9 +126,7 @@ public static void ensurePrincipalAccessIsAllowedForService(final ServiceTicket
     public static void ensurePrincipalAccessIsAllowedForService(final Service service, final RegisteredService registeredService,
                                                                 final TicketGrantingTicket ticketGrantingTicket)
             throws UnauthorizedServiceException, PrincipalException {
-        final List<Authentication> authentications = ticketGrantingTicket.getChainedAuthentications();
-        final Authentication authentication = authentications.get(authentications.size() - 1);
-        ensurePrincipalAccessIsAllowedForService(service, registeredService, authentication);
+        ensurePrincipalAccessIsAllowedForService(service, registeredService, ticketGrantingTicket.getRoot().getAuthentication());
```
Comment (mmoayyed) [inline core/cas-server-core-services/src/main/java/org/apereo/cas/services/RegisteredServiceAccessStrategyUtils.java]: This needs to switch to `getRootAuthenticationFrom` now, right?

File: core/cas-server-core-tickets/src/main/java/org/apereo/cas/ticket/TicketGrantingTicketImpl.java [hunk 2]
```diff
@@ -185,8 +185,7 @@ public synchronized ServiceTicket grantServiceTicket(final String id,
     protected void trackServiceSession(final String id, final Service service, final boolean onlyTrackMostRecentSession) {
         update();
 
-        final List<Authentication> authentications = getChainedAuthentications();
-        service.setPrincipal(authentications.get(authentications.size() - 1).getPrincipal());
+        service.setPrincipal(getRoot().getAuthentication().getPrincipal());
```
Comment (mmoayyed) [inline core/cas-server-core-tickets/src/main/java/org/apereo/cas/ticket/TicketGrantingTicketImpl.java]: This also needs to switch to `getRootAuthenticationFrom` now, right?

File: core/cas-server-core/src/main/java/org/apereo/cas/DefaultCentralAuthenticationService.java [hunk 3]
```diff
@@ -197,8 +196,7 @@ public ProxyTicket grantProxyTicket(final String proxyGrantingTicket, final Serv
         getAuthenticationSatisfiedByPolicy(proxyGrantingTicketObject.getRoot().getAuthentication(),
                 new ServiceContext(service, registeredService));
 
-        final List<Authentication> authentications = proxyGrantingTicketObject.getChainedAuthentications();
-        final Authentication authentication = authentications.get(authentications.size() - 1);
+        final Authentication authentication = proxyGrantingTicketObject.getRoot().getAuthentication();
```
Comment (mmoayyed) [inline core/cas-server-core/src/main/java/org/apereo/cas/DefaultCentralAuthenticationService.java]: How about these?

**Comment by DavidRG13** [comment]
tbh, I don't see any benefit on adding the method to a different class than TGTImpl, adding another import...

tgt.getRoot().getAuthentication() is not that long

up to you mate, let me know, I'll do the change

**Comment by mmoayyed** [comment]
I don't disagree :) which was my original point that `tgt.getRoot().getAuthentication()` and `tgt.getRootAuthentication()` arent really all that different, BUT, if you do wish to add this, it's a support utility that needs go into a support class. I'd be OK with removing it altogether too and simply just use the "long form".

**Comment by mmoayyed** [comment]
Of course, if you do decide to stick with the support utility, then areas that are effected by the change should also be changed to use the new support method. Otherwise, there is no point in adding it!

**Comment by DavidRG13** [comment]
let's use the long form :)

**Comment by mmoayyed** [comment]
Outstanding. Thank you!

**Pull request merged**